File: Sample.java
                          // Group Anagrams
// Time Complexity : O(nk) .. n = number of words, k = length of each string
// Space Complexity : O(n)

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        List<List<String>> list = new ArrayList();
        Map<Double, List<String>> map = new HashMap();
        int[] primes = new int[]{2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67,  71, 73, 79, 83, 89, 97, 101, 103};

        for(String str : strs){
            double mult = 1;
            for(char ch : str.toCharArray())
                mult *= primes[ch - 'a'];
            if(map.containsKey(mult))
                map.get(mult).add(str);
            else {
                map.put(mult, new ArrayList<String>());
                map.get(mult).add(str);
            }
        }

        for(Map.Entry<Double, List<String>> mapElement : map.entrySet())
            list.add(new ArrayList(mapElement.getValue()));
        
        return list;
    }
}


                          // Isomorphic Strings
// Time Complexity : O(n) ... n = length of strings s and t
// Space Complexity : O(n) 

class Solution {
    public boolean isIsomorphic(String s, String t) {
        int len = s.length();
        Map<Character, Character> map = new HashMap();
        Set<Character> set = new HashSet();
        for(int i=0;i<len;i++){
            char schar = s.charAt(i), tchar = t.charAt(i);
            if(map.containsKey(schar)){
                if(map.get(schar) != tchar)
                    return false;
            } else {
                if(set.contains(tchar))
                    return false;
                else {
                    map.put(schar, tchar);
                    set.add(tchar);
                }
            }
        }
        return true;
    }
}

                          // Word Pattern
// Time Complexity : o(n) .. n = number of words in string 'pattern'
// Space Complexity : O(1) ... there are max 26 characters.

class Solution {
    public boolean wordPattern(String pattern, String s) {
        String[] list = s.split(" ");
        int len = pattern.length();
        if(list.length != pattern.length())
            return false;
        Map<Character, String> map = new HashMap();
        Set<String> set = new HashSet();

        for(int i=0;i<len;i++){
            char ch = pattern.charAt(i);
            String word = list[i];
            if(map.containsKey(ch)){
                if(!map.get(ch).equals(word))
                    return false;
            } else {
                if(set.contains(word))
                    return false;
                else {
                    map.put(ch, word);
                    set.add(word);
                }
            }
        }

        return true;
    }
}

File: Sample.py
# Problem 1: Group anagrams -> Time Complexicity: O(n log k), where k is average length of each string
# Update: one may use prime products to further reduce the complexicity to O(n) - yet to try

class Solution:
    
    def groupAnagrams(self, strs: List[str]) -> List[List[str]]:
        
        """
        make use of dictionary to group anagrams using sorted string as key and each items with same sorted values appended to a list
        resulted in a list of groups; each item of subgroup having same characters
        """
        
        
        if len(strs) == 0: return []
        d = {}
        for item in strs:
            sorted_item = "".join(sorted(item))
            if not sorted_item in d.keys():
                d[sorted_item] = []
                d[sorted_item].append(item)
            else:
                d[sorted_item].append(item)
                
        return list(d.values())

# Problem 2: Find if strings are Isomorphic -> Time Complexicity: O(n), optimized from O(n^2) by hashing

class Solution:
    
    def isIsomorphic(self, s: str, t: str) -> bool:
        
        """
        use of hash table (dictionary) subjected to definition of isomorphic strings
        - create empty dictionary and update values 
        - if key is already present, check if value is same, otherwise return "False"
        """
        
        sdict = {}; tdict = {}; result = True
        if len(s) != len(t): result = False
        for i in range(len(s)):
            if s[i] in sdict.keys():
                if sdict[s[i]] != t[i]:  result = False
            else: 
                sdict[s[i]] = t[i]
            if t[i] in tdict.keys():
                if tdict[t[i]] != s[i]:  result = False
            else: 
                tdict[t[i]] = s[i]

        return result

# Problem 3: Word Pattern -> Time Complexicity O(n)

class Solution:
    
    def wordPattern(self, pattern: str, s: str) -> bool:
        
        """
        create 2 dictionaries and match key, value pair for each character coming from given set of words
        """
        
        ssplit = s.split(" ")
        c_dict = {}; w_dict = {}
        if len(ssplit) != len(pattern): return False
        
        for c, w in zip(pattern, ssplit):
            if not c in c_dict:
                if w in w_dict:
                    return False 
                else:
                    c_dict[c] = w
                    w_dict[w] = c
            else:
                if c_dict[c] != w: return False
        
        return True